streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsKafkaClient.java

```diff
@@ -0,0 +1,366 @@
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.streams.StreamsConfig;
+import org.apache.kafka.streams.errors.StreamsException;
+
+import java.io.IOException;
+import java.net.InetSocketAddress;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.List;
+import java.util.LinkedHashMap;
+import java.util.Properties;
+import java.util.Collection;
+import java.util.concurrent.TimeUnit;
+
+public class StreamsKafkaClient {
+
+    private final KafkaClient kafkaClient;
+    private final List<MetricsReporter> reporters;
+    private final StreamsConfig streamsConfig;
+
+    private static final int MAX_INFLIGHT_REQUESTS = 100;
+    private static final long MAX_WAIT_TIME_MS = 30000;
+
+    public StreamsKafkaClient(final StreamsConfig streamsConfig) {
+
+        this.streamsConfig = streamsConfig;
+        final Time time = new SystemTime();
+
+        final Map<String, String> metricTags = new LinkedHashMap<>();
+        metricTags.put("client-id", StreamsConfig.CLIENT_ID_CONFIG);
+
+        final Metadata metadata = new Metadata(streamsConfig.getLong(StreamsConfig.RETRY_BACKOFF_MS_CONFIG), streamsConfig.getLong(StreamsConfig.METADATA_MAX_AGE_CONFIG));
+        final List<InetSocketAddress> addresses = ClientUtils.parseAndValidateAddresses(streamsConfig.getList(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG));
+        metadata.update(Cluster.bootstrap(addresses), time.milliseconds());
+
+        final MetricConfig metricConfig = new MetricConfig().samples(streamsConfig.getInt(CommonClientConfigs.METRICS_NUM_SAMPLES_CONFIG))
+                .timeWindow(streamsConfig.getLong(CommonClientConfigs.METRICS_SAMPLE_WINDOW_MS_CONFIG), TimeUnit.MILLISECONDS)
+                .tags(metricTags);
+        reporters = streamsConfig.getConfiguredInstances(ProducerConfig.METRIC_REPORTER_CLASSES_CONFIG,
+                MetricsReporter.class);
+        // TODO: This should come from the KafkaStream
+        reporters.add(new JmxReporter("kafka.admin"));
+        final Metrics metrics = new Metrics(metricConfig, reporters, time);
+
+        final ChannelBuilder channelBuilder = ClientUtils.createChannelBuilder(streamsConfig.values());
+
+        final Selector selector = new Selector(streamsConfig.getLong(StreamsConfig.CONNECTIONS_MAX_IDLE_MS_CONFIG), metrics, time, "kafka-client", channelBuilder);
+
+        kafkaClient = new NetworkClient(
+                selector,
+                metadata,
+                streamsConfig.getString(StreamsConfig.CLIENT_ID_CONFIG),
+                MAX_INFLIGHT_REQUESTS, // a fixed large enough value will suffice
+                streamsConfig.getLong(StreamsConfig.RECONNECT_BACKOFF_MS_CONFIG),
+                streamsConfig.getInt(StreamsConfig.SEND_BUFFER_CONFIG),
+                streamsConfig.getInt(StreamsConfig.RECEIVE_BUFFER_CONFIG),
+                streamsConfig.getInt(StreamsConfig.REQUEST_TIMEOUT_MS_CONFIG), time);
```

Comment:
Do we want to use the same value for `NetworkClient.poll()` timeout, and as well as the timeout value of `create/deleteTopic` requests? In addition the `MAX_WAIT_TIME_MS` is 30 seconds which is even lower than the default value of timeout, so it is likely that the `client.poll()` will not return even when `MAX_WAIT_TIME_MS` has elapsed.